File: skill-test/projects/sample-cluster-java/src/test/java/cluster/java/Cafe24ApiActorTest.java
package cluster.java;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.testkit.javadsl.TestKit;
import cluster.java.cafe24.Cafe24ApiManagerActor;
import cluster.java.cafe24.Cafe24MetricsSingletonActor;
import cluster.java.cafe24.DummyCafe24Api;
import com.typesafe.config.ConfigFactory;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import scala.concurrent.duration.FiniteDuration;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class Cafe24ApiActorTest {
    private static ActorSystem system;
    private static DummyCafe24Api dummyCafe24Api;

    @BeforeAll
    static void setup() {
        system = ActorSystem.create("Cafe24ApiTestSystem", ConfigFactory.load());
        dummyCafe24Api = new DummyCafe24Api(10, 2);
    }

    @AfterAll
    static void teardown() {
        dummyCafe24Api.close();
        TestKit.shutdownActorSystem(system,
            FiniteDuration.apply(10, TimeUnit.SECONDS), true);
    }

    @Test
    void helloShouldBecomeWorldThroughSafeCaller() {
        new TestKit(system) {{
            ActorRef metrics = system.actorOf(Cafe24MetricsSingletonActor.props(), "cafe24-metrics-test");
            ActorRef manager = system.actorOf(
                Cafe24ApiManagerActor.props(2, dummyCafe24Api, metrics), "cafe24-manager-test");

            manager.tell(new Cafe24ApiManagerActor.ApiRequest("mall-a", "hello"), getRef());
            Cafe24ApiManagerActor.ApiResponse response =
                expectMsgClass(Duration.ofSeconds(5), Cafe24ApiManagerActor.ApiResponse.class);

            assertEquals("mall-a", response.mallId());
            assertEquals("world", response.result());
            assertEquals(200, response.statusCode());
            expectNoMessage(Duration.ofMillis(200));
        }};
    }

    @Test
    void perMallBackpressureShouldKeepBurstCallsSuccessfulAndIsolated() {
        new TestKit(system) {{
            ActorRef metrics = system.actorOf(Cafe24MetricsSingletonActor.props(), "cafe24-metrics-test-2");
            ActorRef manager = system.actorOf(
                Cafe24ApiManagerActor.props(2, dummyCafe24Api, metrics), "cafe24-manager-test-2");

            for (int i = 0; i < 6; i++) {
                manager.tell(new Cafe24ApiManagerActor.ApiRequest("mall-a", "a-" + i), getRef());
                manager.tell(new Cafe24ApiManagerActor.ApiRequest("mall-b", "b-" + i), getRef());
            }

            List<Cafe24ApiManagerActor.ApiResponse> responses = new ArrayList<>();
            for (int i = 0; i < 12; i++) {
                responses.add(expectMsgClass(Duration.ofSeconds(20), Cafe24ApiManagerActor.ApiResponse.class));
            }
            assertTrue(responses.stream().allMatch(r -> r.statusCode() == 200));

            awaitAssert(Duration.ofSeconds(5), Duration.ofMillis(200), () -> {
                metrics.tell(new Cafe24MetricsSingletonActor.GetMallMetrics("mall-a"), getRef());
                Cafe24MetricsSingletonActor.MallMetrics mallA =
                    expectMsgClass(Duration.ofMillis(500), Cafe24MetricsSingletonActor.MallMetrics.class);
                assertEquals(6, mallA.totalCalls());
                assertEquals(0, mallA.throttled429());
                return null;
            });

            awaitAssert(Duration.ofSeconds(5), Duration.ofMillis(200), () -> {
                metrics.tell(new Cafe24MetricsSingletonActor.GetMallMetrics("mall-b"), getRef());
                Cafe24MetricsSingletonActor.MallMetrics mallB =
                    expectMsgClass(Duration.ofMillis(500), Cafe24MetricsSingletonActor.MallMetrics.class);
                assertEquals(6, mallB.totalCalls());
                assertEquals(0, mallB.throttled429());
                return null;
            });
        }};
    }
}

File: skill-test/projects/sample-cluster-kotlin/src/test/kotlin/cluster/kotlin/Cafe24ApiActorTest.kt
package cluster.kotlin

import com.typesafe.config.ConfigFactory
import org.apache.pekko.actor.testkit.typed.javadsl.ActorTestKit
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.time.Duration

class Cafe24ApiActorTest {
    companion object {
        private lateinit var testKit: ActorTestKit
        private lateinit var dummyApi: DummyCafe24Api

        @JvmStatic
        @BeforeAll
        fun setup() {
            testKit = ActorTestKit.create(
                ConfigFactory.parseString(
                    """
                    pekko.actor.provider = "local"
                    pekko.loglevel = "INFO"
                    """.trimIndent()
                ).withFallback(ConfigFactory.load())
            )
            dummyApi = DummyCafe24Api(bucketCapacity = 10, leakRatePerSecond = 2)
        }

        @JvmStatic
        @AfterAll
        fun teardown() {
            dummyApi.close()
            testKit.shutdownTestKit()
        }
    }

    @Test
    fun `hello should become world through safe caller`() {
        val metrics = testKit.spawn(Cafe24MetricsSingletonActor.create(), "cafe24-metrics-test")
        val manager = testKit.spawn(
            Cafe24ApiManagerActor.create(
                perMallMaxRequestsPerSecond = 2,
                dummyApi = dummyApi,
                metricsActor = metrics
            ),
            "cafe24-manager-test"
        )
        val responseProbe = testKit.createTestProbe<Cafe24ApiResponse>()

        manager.tell(Cafe24ApiRequest("mall-a", "hello", responseProbe.ref()))

        val response = responseProbe.expectMessageClass(Cafe24ApiResponse::class.java, Duration.ofSeconds(5))
        kotlin.test.assertEquals("mall-a", response.mallId)
        kotlin.test.assertEquals("world", response.result)
        kotlin.test.assertEquals(200, response.statusCode)
    }

    @Test
    fun `per mall backpressure should keep burst calls successful and isolated`() {
        val metrics = testKit.spawn(Cafe24MetricsSingletonActor.create(), "cafe24-metrics-test-2")
        val manager = testKit.spawn(
            Cafe24ApiManagerActor.create(
                perMallMaxRequestsPerSecond = 2,
                dummyApi = dummyApi,
                metricsActor = metrics
            ),
            "cafe24-manager-test-2"
        )
        val responseProbe = testKit.createTestProbe<Cafe24ApiResponse>()
        val metricsProbe = testKit.createTestProbe<Cafe24MetricsResponse>()

        repeat(6) { index ->
            manager.tell(Cafe24ApiRequest("mall-a", "a-$index", responseProbe.ref()))
            manager.tell(Cafe24ApiRequest("mall-b", "b-$index", responseProbe.ref()))
        }

        val responses = (0 until 12).map {
            responseProbe.expectMessageClass(Cafe24ApiResponse::class.java, Duration.ofSeconds(20))
        }
        kotlin.test.assertTrue(responses.all { it.statusCode == 200 })

        metricsProbe.awaitAssert(Duration.ofSeconds(5), Duration.ofMillis(200)) {
            metrics.tell(GetMallMetrics("mall-a", metricsProbe.ref()))
            val mallA = metricsProbe.expectMessageClass(Cafe24MetricsResponse::class.java, Duration.ofMillis(500))
            kotlin.test.assertEquals(6, mallA.totalCalls)
            kotlin.test.assertEquals(0, mallA.throttled429)
            "mall-a-ok"
        }

        metricsProbe.awaitAssert(Duration.ofSeconds(5), Duration.ofMillis(200)) {
            metrics.tell(GetMallMetrics("mall-b", metricsProbe.ref()))
            val mallB = metricsProbe.expectMessageClass(Cafe24MetricsResponse::class.java, Duration.ofMillis(500))
            kotlin.test.assertEquals(6, mallB.totalCalls)
            kotlin.test.assertEquals(0, mallB.throttled429)
            "mall-b-ok"
        }
    }
}
